plugins/inputs/elasticsearch/elasticsearch.go

```diff
@@ -460,6 +483,87 @@ func (e *Elasticsearch) gatherClusterStats(url string, acc telegraf.Accumulator)
 	return nil
 }
 
+func (e *Elasticsearch) gatherIndicesStats(url string, acc telegraf.Accumulator) error {
+	indicesStats := &struct {
+		Shards  map[string]interface{} `json:"_shards"`
+		All     map[string]interface{} `json:"_all"`
+		Indices map[string]indexStat   `json:"indices"`
+	}{}
+
+	if err := e.gatherJSONData(url, indicesStats); err != nil {
+		return err
+	}
+	now := time.Now()
+
+	// Total Shards Stats
+	shardsStats := map[string]interface{}{}
+	for k, v := range indicesStats.Shards {
+		shardsStats[k] = v
+	}
+	acc.AddFields("elasticsearch_indicestats_shards_total", shardsStats, map[string]string{"name": ""}, now)
+
+	// All Stats
+	for m, s := range indicesStats.All {
+		// parse Json, ignoring strings and bools
+		jsonParser := jsonparser.JSONFlattener{}
+		err := jsonParser.FullFlattenJSON("_", s, true, true)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("elasticsearch_indicestats_"+m, jsonParser.Fields, map[string]string{"index_name": "all"}, now)
+	}
+
+	// Individual Indices stats
+	for id, index := range indicesStats.Indices {
+		indexTag := map[string]string{"index_name": id}
+		stats := map[string]interface{}{
+			"primaries": index.Primaries,
+			"total":     index.Total,
+		}
+		for m, s := range stats {
+			f := jsonparser.JSONFlattener{}
+			// parse Json, getting strings and bools
+			err := f.FullFlattenJSON("", s, true, true)
+			if err != nil {
+				return err
+			}
+			acc.AddFields("elasticsearch_indicestats_"+m, f.Fields, indexTag, now)
+		}
+
+		if e.ShardsStats {
+			for shardNumber, shard := range index.Shards {
+
+				// Get Shard Stats
+				flattened := jsonparser.JSONFlattener{}
+				err := flattened.FullFlattenJSON("", shard[0], true, true)
+				if err != nil {
+					return err
+				}
+
+				// determine shard tag and primary/replica designation
+				shardType := "replica"
+				if flattened.Fields["routing_primary"] == true {
```

Comment:
We should remove this as a field so that it only appears once in the metric.  Not directly related to this line, but I think it also makes sense to remove some select items from the output.  For example `shard_path_data_path` and `shard_path_state_path` don't seem very useful, and I doubt we want to collect any random string values.

Reply:
The more we try to filter, the more changes will be needed to the code I think.
Cause every version of ElasticSearch has some other/additional fields.

Don't think its a big issue to have 'to much' data?

Or we should add a []string in the config where you can specify the things you want to ignore...

Reply:
Actually I think there is a bit of risk here around having too much data, and it does push some responsibility to other users to fielddrop these fields.  The fielddrop/fieldpass would be the way to let the user blacklist the data, I don't think we need an option here explicitly for it.

Since this is a dynamic plugin, I'd say we should blacklist all strings now, and if we feel like specific ones ought to be included we can whitelist them back in.

Reply:
Any suggestions how to properly blacklist the strings?

Reply:
Third argument to FullFlattenJSON should disable strings.

Reply:
But then I would need to have the FullFlattenJSON twice ... As I need some strings to use as a key (routing_node for example).
Or you have a better option? :)
Looping at the end over all the fields to remove the strings doesn't feel like a clean solution :)

Reply:
Good point but I think looping over the fields an removing the strings is probably what we should do.

Alternately, but I think maybe you won't prefer this, we could create struct types for the response and use pointers for the fields which may or may not exist.  Anything we don't want we leave off.  This gives the most control over the output which can be helpful to avoid issues where the type changes.

Reply:
Resolved since last commit.
Just added an additional option to convert routing_state to an int, cause thats quite important information to report :)